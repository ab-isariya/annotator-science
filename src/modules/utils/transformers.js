import DateTime from 'luxon/src/datetime';

/**
 * Returns our UTC datetimes in a given format
 *
 * @param {String} datetimeStr
 * @param {Object} format - valid object to Luxon
 * https://moment.github.io/luxon/docs/class/src/datetime.js~DateTime.html
 * @return {string}
 */
export function datetimeToHuman(datetimeStr, format) {
  return DateTime.fromISO(datetimeStr, {zone: 'utc'})
    .toLocal()
    .toLocaleString(format);
}

/**
 * Returns the current datetime as an ISO.
 *
 * NOTE! It removes the Z from the end because AWS RDS doesn't like keeping the Z
 * in it's records.
 *
 * @return {string}
 */
export function nowISOTime() {
  return new Date().toISOString().slice().slice(0, -1); //NOTE(Rejon): Drop the Z!;
}

/**
 * Returns size in human readable units
 *
 * @param {Number} sizeInBytes - size in bytes
 * @return {String} human readable size and the unit
 */
export function humanFileSize(sizeInBytes) {
  const i =
    sizeInBytes === 0 ? 0 : Math.floor(Math.log(sizeInBytes) / Math.log(1024));
  return (
    (sizeInBytes / Math.pow(1024, i)).toFixed(1) +
    ['B', 'kB', 'MB', 'GB', 'TB'][i]
  );
}

/**
 * Rounding function that avoids the problem of decimal of 5 rounded down
 * instead of up
 *
 * @param {Number} value
 * @param {Number} decimals
 * @return {number}
 */
export function round(value, decimals) {
  return Number(Math.round(value + 'e' + decimals) + 'e-' + decimals);
}

/**
 * Formats a number (integer or float) into a string with comma
 * separators. Also handles decimals for floats.

 * @param {number} amount - Number to be formatted
 * @return {string}
 */
export const CommaFormatted = (amount) => {
  var delimiter = ','; // replace comma if desired
  var a = String(amount).split('.', 2);
  var d = a[1];
  var i = parseInt(a[0]);
  if (isNaN(i)) {
    return '';
  }
  var minus = '';
  if (i < 0) {
    minus = '-';
  }
  i = Math.abs(i);
  var n = String(i);
  a = [];
  while (n.length > 3) {
    var nn = n.substr(n.length - 3);
    a.unshift(nn);
    n = n.substr(0, n.length - 3);
  }
  if (n.length > 0) {
    a.unshift(n);
  }
  n = a.join(delimiter);
  if (d === undefined || d.length < 1) {
    amount = n;
  } else {
    amount = n + '.' + d;
  }
  amount = minus + amount;
  return amount;
};

/**
 * Formats a number to a shortend place (ie. 1,000 becomes 1.0k)
 * //From: https://stackoverflow.com/questions/9461621/format-a-number-as-2-5k-if-a-thousand-or-more-otherwise-900
 * @constructor
 * @param {number} num - Number to be formatted.
 * @param {number} digits - Amount of digits to support when formatting (ie. 10,080 with 3 digits -> 10.1k with 4 digits -> 10.05k)
 * @return {string}
 */
export const nFormatter = (num, digits) => {
  var si = [
    {value: 1, symbol: ''},
    {value: 1e3, symbol: 'k'},
    {value: 1e6, symbol: 'M'},
    {value: 1e9, symbol: 'G'},
    {value: 1e12, symbol: 'T'},
    {value: 1e15, symbol: 'P'},
    {value: 1e18, symbol: 'E'}
  ];
  var rx = /\.0+$|(\.[0-9]*[1-9])0+$/;
  var i;
  for (i = si.length - 1; i > 0; i--) {
    if (num >= si[i].value) {
      break;
    }
  }
  return (num / si[i].value).toFixed(digits).replace(rx, '$1') + si[i].symbol;
};

/**
 * Formatter that combines nFormatter and CommaFormatted
 * based on a limit. If num is less than the limit, it'll be comma formatted,
 * else, it'll be nFormatted.
 * NOTE: limit is 1000 by default
 *
 * @param {number} num - Number to be formatted.
 * @param {number} digits - Number of digits to show for nFormatted
 * @param {number} limit (Default: 1000) - Limit to change the format from CommaSeperated into nFormatted.
 * @return {string}
 */
export const NumberCommaSuffix = (num, digits, limit = 1000) => {
  if (isNaN(num)) {
    return '';
  }

  if (num <= limit) {
    return CommaFormatted(String(num));
  } else {
    return nFormatter(num, digits);
  }
};
