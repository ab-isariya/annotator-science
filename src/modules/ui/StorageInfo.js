import React from 'react';
import PropTypes from 'prop-types';

import styled from 'styled-components';

import Button from '@modules/ui/Button';
import {humanFileSize} from '@utils/transformers';

import {ReactComponent as Storage} from '@assets/svgs/Storage.svg';

const MeterContainer = styled.div`
  height: 16px;
`;

const MeterDiv = styled.div`
  background: rgb(169, 185, 236);
  background: linear-gradient(
    18deg,
    rgba(169, 185, 236, 1) 0%,
    rgba(152, 147, 247, 1) 16%,
    rgba(118, 135, 242, 1) 30%,
    rgba(83, 139, 231, 1) 44%,
    rgba(75, 158, 218, 1) 58%,
    rgba(124, 198, 199, 1) 100%
  );
  width: ${(props) => `${props.size}%`};
`;

/**
 * Displays storage used by user
 *
 * @param {Number} used - space used by user, in bytes
 * @param {Number} total - total space available to user, in bytes
 */
const StorageInfo = ({used, total}) => {
  const meterWidthPercentage = (100 * used) / total;

  return (
    <div className="m-2.5 p-4 bg-white rounded-md">
      <div className="flex items-center space-x-2.5 font-inter font-light text-base leading-6">
        <Storage />
        <div>Storage</div>
      </div>

      <div className="mt-2.5 mb-1 font-inter font-normal text-sm leading-5 text-blue-500">
        <b className="font-medium">{humanFileSize(used)}</b> of{' '}
        {humanFileSize(total)}
      </div>

      <MeterContainer className="relative bg-grey-100 rounded-md my-0.5">
        <MeterDiv
          className="h-full relative overflow-hidden rounded-md"
          size={meterWidthPercentage}>
          &nbsp;
        </MeterDiv>
      </MeterContainer>

      <p className="mb-4 font-inter font-normal text-xs leading-4">
        5Mb/doc upload limit
      </p>

      <Button secondary fullWidth>
        Upgrade
      </Button>
    </div>
  );
};

StorageInfo.propTypes = {
  /** space used by user, in bytes */
  used: PropTypes.number.isRequired,
  /** total space available to user, in bytes */
  total: PropTypes.number.isRequired
};

export default StorageInfo;
