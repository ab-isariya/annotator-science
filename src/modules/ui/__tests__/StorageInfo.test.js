import StorageInfo from '../StorageInfo';
import {humanFileSize} from '@utils/transformers';

import renderer from 'react-test-renderer';

describe('<StorageInfo/>', () => {
  it('Renders without crashing', () => {
    const comp = renderer
      .create(<StorageInfo used={100} total={150} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
  });

  it('Renders with a positive percentage', () => {
    const comp = renderer
      .create(<StorageInfo used={25} total={100} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
  });

  it('Renders with a negative percentage', () => {
    const comp = renderer
      .create(<StorageInfo used={-25} total={100} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
  });

  it('Renders with a 100% total', () => {
    const comp = renderer
      .create(<StorageInfo used={100} total={100} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
  });

  it('Renders with a total greater than 100%', () => {
    let fileSizeMax = humanFileSize(110);

    const comp = renderer
      .create(<StorageInfo used={110} total={100} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
    expect(comp.children[1].children[0].children[0]).toBe(fileSizeMax);
  });
});
