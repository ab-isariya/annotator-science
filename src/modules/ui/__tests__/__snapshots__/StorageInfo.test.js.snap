// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StorageInfo/> Renders with a 100% total 1`] = `
<div
  className="m-2.5 p-4 bg-white rounded-md"
>
  <div
    className="flex items-center space-x-2.5 font-inter font-light text-base leading-6"
  >
    <svg>
      Storage.svg
    </svg>
    <div>
      Storage
    </div>
  </div>
  <div
    className="mt-2.5 mb-1 font-inter font-normal text-sm leading-5 text-blue-500"
  >
    <b
      className="font-medium"
    >
      100.0B
    </b>
     of
     
    100.0B
  </div>
  <div
    className="sc-bdvvtL dxGirE relative bg-grey-100 rounded-md my-0.5"
  >
    <div
      className="sc-gsDKAQ jmTlin h-full relative overflow-hidden rounded-md"
      size={100}
    >
       
    </div>
  </div>
  <p
    className="mb-4 font-inter font-normal text-xs leading-4"
  >
    5Mb/doc upload limit
  </p>
  <button
    className="filter rounded-md font-inter transition-all outline-none  bg-white border border-grey-500 hover:bg-grey-50 focus:bg-grey-50 focus:drop-shadow-focus-button flex flex-row items-center gap-x-2 text-sm p-2.5"
    fullWidth={true}
  >
    Upgrade
  </button>
</div>
`;

exports[`<StorageInfo/> Renders with a negative percentage 1`] = `
<div
  className="m-2.5 p-4 bg-white rounded-md"
>
  <div
    className="flex items-center space-x-2.5 font-inter font-light text-base leading-6"
  >
    <svg>
      Storage.svg
    </svg>
    <div>
      Storage
    </div>
  </div>
  <div
    className="mt-2.5 mb-1 font-inter font-normal text-sm leading-5 text-blue-500"
  >
    <b
      className="font-medium"
    >
      NaNundefined
    </b>
     of
     
    100.0B
  </div>
  <div
    className="sc-bdvvtL dxGirE relative bg-grey-100 rounded-md my-0.5"
  >
    <div
      className="sc-gsDKAQ iWdJco h-full relative overflow-hidden rounded-md"
      size={-25}
    >
       
    </div>
  </div>
  <p
    className="mb-4 font-inter font-normal text-xs leading-4"
  >
    5Mb/doc upload limit
  </p>
  <button
    className="filter rounded-md font-inter transition-all outline-none  bg-white border border-grey-500 hover:bg-grey-50 focus:bg-grey-50 focus:drop-shadow-focus-button flex flex-row items-center gap-x-2 text-sm p-2.5"
    fullWidth={true}
  >
    Upgrade
  </button>
</div>
`;

exports[`<StorageInfo/> Renders with a positive percentage 1`] = `
<div
  className="m-2.5 p-4 bg-white rounded-md"
>
  <div
    className="flex items-center space-x-2.5 font-inter font-light text-base leading-6"
  >
    <svg>
      Storage.svg
    </svg>
    <div>
      Storage
    </div>
  </div>
  <div
    className="mt-2.5 mb-1 font-inter font-normal text-sm leading-5 text-blue-500"
  >
    <b
      className="font-medium"
    >
      25.0B
    </b>
     of
     
    100.0B
  </div>
  <div
    className="sc-bdvvtL dxGirE relative bg-grey-100 rounded-md my-0.5"
  >
    <div
      className="sc-gsDKAQ iReeWR h-full relative overflow-hidden rounded-md"
      size={25}
    >
       
    </div>
  </div>
  <p
    className="mb-4 font-inter font-normal text-xs leading-4"
  >
    5Mb/doc upload limit
  </p>
  <button
    className="filter rounded-md font-inter transition-all outline-none  bg-white border border-grey-500 hover:bg-grey-50 focus:bg-grey-50 focus:drop-shadow-focus-button flex flex-row items-center gap-x-2 text-sm p-2.5"
    fullWidth={true}
  >
    Upgrade
  </button>
</div>
`;

exports[`<StorageInfo/> Renders with a total greater than 100% 1`] = `
<div
  className="m-2.5 p-4 bg-white rounded-md"
>
  <div
    className="flex items-center space-x-2.5 font-inter font-light text-base leading-6"
  >
    <svg>
      Storage.svg
    </svg>
    <div>
      Storage
    </div>
  </div>
  <div
    className="mt-2.5 mb-1 font-inter font-normal text-sm leading-5 text-blue-500"
  >
    <b
      className="font-medium"
    >
      110.0B
    </b>
     of
     
    100.0B
  </div>
  <div
    className="sc-bdvvtL dxGirE relative bg-grey-100 rounded-md my-0.5"
  >
    <div
      className="sc-gsDKAQ jnbHMu h-full relative overflow-hidden rounded-md"
      size={110}
    >
       
    </div>
  </div>
  <p
    className="mb-4 font-inter font-normal text-xs leading-4"
  >
    5Mb/doc upload limit
  </p>
  <button
    className="filter rounded-md font-inter transition-all outline-none  bg-white border border-grey-500 hover:bg-grey-50 focus:bg-grey-50 focus:drop-shadow-focus-button flex flex-row items-center gap-x-2 text-sm p-2.5"
    fullWidth={true}
  >
    Upgrade
  </button>
</div>
`;

exports[`<StorageInfo/> Renders without crashing 1`] = `
<div
  className="m-2.5 p-4 bg-white rounded-md"
>
  <div
    className="flex items-center space-x-2.5 font-inter font-light text-base leading-6"
  >
    <svg>
      Storage.svg
    </svg>
    <div>
      Storage
    </div>
  </div>
  <div
    className="mt-2.5 mb-1 font-inter font-normal text-sm leading-5 text-blue-500"
  >
    <b
      className="font-medium"
    >
      100.0B
    </b>
     of
     
    150.0B
  </div>
  <div
    className="sc-bdvvtL dxGirE relative bg-grey-100 rounded-md my-0.5"
  >
    <div
      className="sc-gsDKAQ bByECZ h-full relative overflow-hidden rounded-md"
      size={66.66666666666667}
    >
       
    </div>
  </div>
  <p
    className="mb-4 font-inter font-normal text-xs leading-4"
  >
    5Mb/doc upload limit
  </p>
  <button
    className="filter rounded-md font-inter transition-all outline-none  bg-white border border-grey-500 hover:bg-grey-50 focus:bg-grey-50 focus:drop-shadow-focus-button flex flex-row items-center gap-x-2 text-sm p-2.5"
    fullWidth={true}
  >
    Upgrade
  </button>
</div>
`;
